Include/internal/pycore_interp.h

```diff
@@ -39,6 +39,22 @@ struct _Py_long_state {
     int max_str_digits;
 };
 
+// Support for stop-the-world events. This exists in both the PyRuntime struct
+// for global pauses and in each PyInterpreterState for per-interpreter pauses.
+struct _stoptheworld_state {
+    PyMutex mutex;       // Serializes stop-the-world attempts.
+
+    // NOTE: The fields below are protected by HEAD_LOCK(runtime), not by the
+    // above mutex.
+    bool requested;      // Set when a pause is requested.
+    bool world_stopped;  // Set when the world is stopped.
+    bool is_global;      // Set when contained in PyRuntime struct.
+
+    PyEvent stop_event;  // Set when thread_countdown reaches zero.
+    Py_ssize_t thread_countdown;  // Number of threads that must pause.
+
+    PyThreadState *requester; // Thread that requested the pause (may be NULL).
+};
```

Comment:
Thanks for explaining.  It still feels like a ceval feature to me though.  Here's why:

* (as noted) it is executed via the eval breaker
* it has the same vibe as pretty much all the other things you find in Python/ceval_gil.c
* it is a replacement for functionality that the GIL provides currently (and the GIL is a ceval feature)

For a specific example, it feels kind of similar to `Py_AddPendingCall()`, which you'll find defined/implemented in ceval-related files.

As a counterpoint, the main motivation for stop-the-world is to facilitate GC.  GC is generally effected through the eval breaker, but GC isn't defined in ceval-related files; it *is* in its own feature-specific files.  (Stop-the-world doesn't feel like something that should have it's own files.)

Another counter example: stop-the-world definitely has a runtime lifecycle vibe too.  I'm reminded of `PyInterpreterState._finalizing` and `_PyRuntimeState._finalizing` (which both incidentally relate to a single thread state).  However, stop-the-world is all about thread states in an interpreter or the global runtime, rather than about those two state structs themselves, nor about their lifecycle.

> It's fundamentally about `PyThreadState.state`: transitioning threads to the _Py_THREAD_GC state.

Yeah, that makes sense.  Perhaps it should go in pycore_tstate.h.

Regardless, in the details this is certainly about more than transitioning `PyThreadState.state`.  It also includes actually pausing/unpausing threads, as well as establishing the state *all* threads should be in if any is created or tries to transition why the umbrella meta-state is in effect.

Finally, if I had to associate stop-the-world with a single feature, it would be code execution (specifically, synchronizing it, a la the GIL).  To me, "code execution" equates to ceval.

Reply:
FWIW, one thing that threw me off is that this would be the only case where we define a struct in pycore_interp.h that we also use in pycore_runtime.h (or vice versa).  There are certainly cases where we use use the same struct in both, but such structs are all defined in feature-specific header files.